### PR TITLE
Iss6

### LINKED ISSUE #3: Clear (and/or refresh) BugPathList when necessary
Right now BugPathList does not get cleared when switching to another file or refreshing the currently selected file (refreshing by unfocusing/focusing) after a rebuild which possibly removes the report that has its BugPath in the BugPathList.

### PR DESCRIPTION
**Pull request by Locutus18**
Fixed #3 and #6. 
And fixes the fact that if the same file editor clicks than what you do not upgrade your current or can not close to the ReportList.

### CODE REVIEW

File: eclipse-plugin/eclipse/cc.codechecker.eclipse.plugin/src/cc/codechecker/plugin/config/CodeCheckerContext.java [hunk 1]
```diff
@@ -56,9 +56,9 @@
 
     private CodeCheckerContext() {
         PerServerActionRunner<ThriftCommunicationInterface> actionRunner;
-        actionRunner = new PerServerSimpleActionRunner<ThriftCommunicationInterface>(new
-                ThriftTransportFactory(), (new CodecheckerActionInitializer()).initialize(new
-                ActionImplementationRegistry<ThriftCommunicationInterface>()));
+        actionRunner = new PerServerSimpleActionRunner<ThriftCommunicationInterface>(new ThriftTransportFactory(),
```
Comment (dutow) [inline eclipse-plugin/eclipse/cc.codechecker.eclipse.plugin/src/cc/codechecker/plugin/config/CodeCheckerContext.java]: it's usually a good idea to don't change the formatting (or do so in a separate commit), as it can make merging more complicated. Especially for files totally unrelated to the commit message.

File: eclipse-plugin/eclipse/cc.codechecker.eclipse.plugin/src/cc/codechecker/plugin/views/report/details/BugPathContentProvider.java [hunk 2]
```diff
@@ -24,7 +27,13 @@ public void inputChanged(Viewer viewer, Object oldInput, Object newInput) {
...
                     inputElement;
             if (bp.isPresent()) {
-                return bp.get().getItems().toArray();
+                ArrayList<BugPathItem> result = new ArrayList<>(bp.get().getItems().asList());
+                for (int i = result.size() - 1; i >= 0; --i) {
+                    if (result.get(i).getMessage() == "") {
+                        result.remove(i);
+                    }
+                }
+                return result.toArray();
```
Comment (dutow) [inline eclipse-plugin/eclipse/cc.codechecker.eclipse.plugin/src/cc/codechecker/plugin/views/report/details/BugPathContentProvider.java]: Guava contains a function named [removeIf](https://google.github.io/guava/releases/snapshot/api/docs/com/google/common/collect/Iterables.html#removeIf(java.lang.Iterable,%20com.google.common.base.Predicate)) for exactly this loop.

File: eclipse-plugin/eclipse/cc.codechecker.eclipse.plugin/src/cc/codechecker/plugin/config/CodeCheckerContext.java [hunk 3]
```diff
@@ -117,12 +119,13 @@ public void setActiveEditorPart(IEditorPart partRef) {
         if (!(partRef.getEditorInput() instanceof IFileEditorInput)) {
             return;
         }
-
+        if (partRef == activeEditorPart) {
+            return;
+        }
```
Comment (dutow) [inline eclipse-plugin/eclipse/cc.codechecker.eclipse.plugin/src/cc/codechecker/plugin/config/CodeCheckerContext.java]: Are you sure this won't cause any side effects, for example after building the project? This early return means the report list won't be refreshed when the active editor is the same as before, and that's probably not what's expected from this method. Maybe this condition would be better at the inner if for the BugPathListView?

File: eclipse-plugin/eclipse/cc.codechecker.eclipse.plugin/src/cc/codechecker/plugin/views/report/details/BugPathLabelProvider.java [hunk 4]
```diff
@@ -2,17 +2,22 @@
...
     @Override
     public String getText(Object element) {
-        
+
         BugPathItem bpi = (BugPathItem) element;
         Path path = new Path(bpi.getFile());
-        System.out.println(path);
+
+        IPath path2 = ResourcesPlugin.getWorkspace().getRoot().getFile(path).getProjectRelativePath();
+        System.out.println(path2);
```
Comment (dutow) [inline eclipse-plugin/eclipse/cc.codechecker.eclipse.plugin/src/cc/codechecker/plugin/views/report/details/BugPathLabelProvider.java]: This variable is only used for (debug) printing, is it intended?

File: eclipse-plugin/eclipse/cc.codechecker.eclipse.plugin/src/cc/codechecker/plugin/views/report/details/BugPathListView.java [hunk 5]
```diff
@@ -72,8 +72,11 @@ public void setName(String value) {
         setPartName(value);
     }
 
-    public void clear() {
-        viewer.setInput(null);
+    public void clean(boolean delegate) {
+        if(viewer != null && !delegate) {
+            viewer.setInput(null);
+            viewer.refresh();
+        }
```
Comment (dutow) [inline eclipse-plugin/eclipse/cc.codechecker.eclipse.plugin/src/cc/codechecker/plugin/views/report/details/BugPathListView.java]: I don't understand this. If delegate (why is it called delegate?) is false, this method won't do anything. The condition should be outside this function.

Comment (Locutus18) [inline eclipse-plugin/eclipse/cc.codechecker.eclipse.plugin/src/cc/codechecker/plugin/views/report/details/BugPathListView.java]: I have introduced this extra variable to make sure the function is called by the good thread. This function is called by a function which is called from two distinct locations. One of the callers are running on the GUI thread which has authority over this element, so it's fine. But the other caller is on a separate thread, so its call effectively crashes because of cross-thread action.

Comment (dutow) [inline eclipse-plugin/eclipse/cc.codechecker.eclipse.plugin/src/cc/codechecker/plugin/views/report/details/BugPathListView.java]: I understand the reason for cleaning the viewer only on certain conditions - but I still can't see why we handle this condition (which is unrelated to clearing the view) inside the clear function. You said now that the reason behind the condition is a different call-stack / thread, which isn't visible at all in this condition.

Think about writing a doc comment for this function (or write one for it). The intent and side effect of a function named clean() is easy to guess and clear. A function named clear(boolean) is already harder to understand. 
And if I read that it actually won't do anything when given a true parameter, I would start second guessing the API docs.

File: eclipse-plugin/eclipse/cc.codechecker.eclipse.plugin/src/cc/codechecker/plugin/views/report/details/BugPathLabelProvider.java (hunk 4 above)
Comment (Locutus18) [inline eclipse-plugin/eclipse/cc.codechecker.eclipse.plugin/src/cc/codechecker/plugin/views/report/details/BugPathLabelProvider.java]: Complete!

File: eclipse-plugin/eclipse/cc.codechecker.eclipse.plugin/src/cc/codechecker/plugin/config/CodeCheckerContext.java (hunk 3 above)
Comment (Locutus18) [inline eclipse-plugin/eclipse/cc.codechecker.eclipse.plugin/src/cc/codechecker/plugin/config/CodeCheckerContext.java]: Complete, plus parameters with function which is responsible for knowing that we are post building the project.

File: eclipse-plugin/eclipse/cc.codechecker.eclipse.plugin/src/cc/codechecker/plugin/views/report/details/BugPathContentProvider.java (hunk 2 above)
Comment (Locutus18) [inline eclipse-plugin/eclipse/cc.codechecker.eclipse.plugin/src/cc/codechecker/plugin/views/report/details/BugPathContentProvider.java]: Complete.

File: eclipse-plugin/eclipse/cc.codechecker.eclipse.plugin/src/cc/codechecker/plugin/config/CodeCheckerContext.java (hunk 1 above)
Comment (Locutus18) [inline eclipse-plugin/eclipse/cc.codechecker.eclipse.plugin/src/cc/codechecker/plugin/config/CodeCheckerContext.java]: Deleted to commit.
